File: src/record/zoneGuard.ts
import { hrZoneIndex } from '@/metrics/deriveSummary';
import type { HrZone } from '@/model/intervalPlan';

export type ZoneStatus = 'in' | 'high' | 'low' | 'none';
export type ZoneEvent = 'high' | 'low' | 'back';

const OUT_OF_ZONE_ALERT_MS = 15_000; // must be continuously out this long before alerting
const REPEAT_ALERT_MS = 60_000; // re-alert at most this often while still out

/**
 * Tracks whether the live HR is in/above/below a target zone and decides when to fire
 * an alert. Feed it a sample every tick (or call `reset()` on target/step change, or
 * while paused). Pure state machine — no timers, no I/O; call `sample()` with the
 * current wall-clock ms and current HR.
 */
export class ZoneGuard {
  private status: ZoneStatus = 'none';
  private outSinceMs: number | null = null; // when the current out-of-zone streak started
  private lastAlertMs: number | null = null; // last time an alert (high/low) fired
  private alerted = false; // an alert fired for the current out-of-zone streak (arms 'back')
  private outSide: ZoneStatus | null = null; // 'high' | 'low' of the current streak

  /** Current in/high/low/none status, for UI coloring. */
  getStatus(): ZoneStatus {
    return this.status;
  }

  /** Target/step changed (or paused/resumed): drop all timers and status. */
  reset(): void {
    this.status = 'none';
    this.outSinceMs = null;
    this.lastAlertMs = null;
    this.alerted = false;
    this.outSide = null;
  }

  /**
   * Feed one sample. `hr`/`targetZone` null, or `paused`, means "ignore" — timers hold
   * but nothing fires (caller should `reset()` on an actual step/target change).
   * Returns the event to fire, if any.
   */
  sample(nowMs: number, hr: number | null, targetZone: HrZone | null, maxHr: number | null, paused = false): ZoneEvent | null {
    if (paused) {
      // Pause time must not count as out-of-zone time (resume would alert at once).
      this.outSinceMs = null;
      this.outSide = null;
      return null;
    }
    if (!hr || !targetZone || !maxHr) return null;

    const zoneIdx = hrZoneIndex(hr, maxHr); // 0-based
    const status: ZoneStatus = zoneIdx + 1 === targetZone ? 'in' : zoneIdx + 1 > targetZone ? 'high' : 'low';
    this.status = status;

    if (status === 'in') {
      const wasAlerted = this.alerted;
      this.outSinceMs = null;
      this.alerted = false;
      if (wasAlerted) return 'back';
      return null;
    }

    // Out of zone (high or low).
    // New streak, or flipped high↔low: restart the 15 s wait; the other side's
    // repeat window doesn't apply (it's a different instruction).
    if (this.outSinceMs == null || this.outSide !== status) {
      if (this.outSide != null && this.outSide !== status) this.lastAlertMs = null;
      this.outSinceMs = nowMs;
      this.outSide = status;
    }
    const outFor = nowMs - this.outSinceMs;
    if (outFor < OUT_OF_ZONE_ALERT_MS) return null;
    if (this.lastAlertMs != null && nowMs - this.lastAlertMs < REPEAT_ALERT_MS) return null;

    this.lastAlertMs = nowMs;
    this.alerted = true;
    return status; // 'high' | 'low'
  }
}

File: android/wear/src/main/java/com/aera/app/wear/ZoneGuard.kt
package com.aera.app.wear

/** in/high/low relative to the target zone, or none while there's no target/reading. */
enum class ZoneStatus { IN, HIGH, LOW, NONE }

/** Alert to fire on a `sample()` call, or null. */
enum class ZoneEvent { HIGH, LOW, BACK }

private const val OUT_OF_ZONE_ALERT_MS = 15_000L
private const val REPEAT_ALERT_MS = 60_000L

/**
 * Watch-side mirror of src/record/zoneGuard.ts — same state machine, same thresholds.
 * Pure Kotlin, no Android deps, so it's unit-testable like PlanRunner.
 */
class ZoneGuard {
    var status: ZoneStatus = ZoneStatus.NONE
        private set
    private var outSinceMs: Long? = null
    private var lastAlertMs: Long? = null
    private var alerted = false
    private var outSide: ZoneStatus? = null // HIGH or LOW of the current streak

    fun reset() {
        status = ZoneStatus.NONE
        outSinceMs = null
        lastAlertMs = null
        alerted = false
        outSide = null
    }

    /** targetZone: 1..5, or null/0 for "no target". */
    fun sample(nowMs: Long, hr: Int, targetZone: Int?, maxHr: Int, paused: Boolean = false): ZoneEvent? {
        if (paused) {
            // Pause time must not count as out-of-zone time (resume would alert at once).
            outSinceMs = null
            outSide = null
            return null
        }
        if (hr <= 0 || targetZone == null || targetZone <= 0 || maxHr <= 0) return null

        val zoneIdx = Zones.index(hr, maxHr) // 0-based
        val s = when {
            zoneIdx + 1 == targetZone -> ZoneStatus.IN
            zoneIdx + 1 > targetZone -> ZoneStatus.HIGH
            else -> ZoneStatus.LOW
        }
        status = s

        if (s == ZoneStatus.IN) {
            val wasAlerted = alerted
            outSinceMs = null
            alerted = false
            return if (wasAlerted) ZoneEvent.BACK else null
        }

        // New streak, or flipped high<->low: restart the 15 s wait; the other side's
        // repeat window doesn't apply (it's a different instruction).
        if (outSinceMs == null || outSide != s) {
            if (outSide != null && outSide != s) lastAlertMs = null
            outSinceMs = nowMs
            outSide = s
        }
        val outFor = nowMs - outSinceMs!!
        if (outFor < OUT_OF_ZONE_ALERT_MS) return null
        val last = lastAlertMs
        if (last != null && nowMs - last < REPEAT_ALERT_MS) return null

        lastAlertMs = nowMs
        alerted = true
        return if (s == ZoneStatus.HIGH) ZoneEvent.HIGH else ZoneEvent.LOW
    }
}
